File: python/tests/test_version_adapter_isolation.py
"""Structural guard: no File_Version adapter may depend on another's.

Each `capabilities/vX_Y/` adapter (and the test-only mock v1.1 adapter in
`test_adapter_migration.py`, which models what a real adapter looks like)
must be fully self-contained. If v1.1 depends on v1.0's code, v1.0 can never
change or be removed later without checking v1.1 — and every later version
compounds the risk. See V1_1_IMPLEMENTATION_PLAN.md's "second guardrail" for
the full reasoning; this test exists so that violation can't land silently
again the way it did once already this session.

The only legitimate place two versions are referenced together is the three
shared dispatch registries (`reader.py`, `writer.py`,
`capabilities/__init__.py`) — this test does not scan those, only the
version-specific adapter packages and the mock.
"""
from __future__ import annotations

import ast
import re
from pathlib import Path

_REPO_ROOT = Path(__file__).parent.parent.parent
_CAPABILITIES_DIR = _REPO_ROOT / "python" / "src" / "machine_config" / "capabilities"
_MOCK_FILE = _REPO_ROOT / "python" / "tests" / "test_adapter_migration.py"

_VERSION_DIR_RE = re.compile(r"^v(\d+)_(\d+)$")
_VERSION_TOKEN_RE = re.compile(r"\bv(\d+)_(\d+)\b")


def _version_dirs() -> list[Path]:
    return sorted(
        p for p in _CAPABILITIES_DIR.iterdir()
        if p.is_dir() and _VERSION_DIR_RE.match(p.name)
    )


def _import_tokens(py_file: Path) -> list[str]:
    """All dotted names/module strings referenced by import statements."""
    tree = ast.parse(py_file.read_text(encoding="utf-8"), filename=str(py_file))
    tokens: list[str] = []
    for node in ast.walk(tree):
        if isinstance(node, ast.Import):
            tokens.extend(alias.name for alias in node.names)
        elif isinstance(node, ast.ImportFrom):
            if node.module:
                tokens.append(node.module)
            tokens.extend(alias.name for alias in node.names)
    return tokens


def _violations_in_file(py_file: Path, own_version: str) -> list[str]:
    violations = []
    for token in _import_tokens(py_file):
        for match in _VERSION_TOKEN_RE.finditer(token):
            found_version = f"v{match.group(1)}_{match.group(2)}"
            if found_version != own_version:
                violations.append(
                    f"{py_file.relative_to(_REPO_ROOT)}: imports '{token}' — "
                    f"references {found_version} from within {own_version}"
                )
    return violations


def test_no_version_adapter_depends_on_another():
    """Nothing under capabilities/vX_Y/ may reference capabilities/vZ_W/ for
    a different (Z, W) — each version's adapter must be independently
    hand-authored and self-contained.
    """
    version_dirs = _version_dirs()
    assert len(version_dirs) >= 1, "expected at least one capabilities/vX_Y directory"

    violations: list[str] = []
    for vdir in version_dirs:
        own_version = vdir.name  # e.g. "v1_0"
        for py_file in vdir.rglob("*.py"):
            violations.extend(_violations_in_file(py_file, own_version))

    assert not violations, (
        "Version adapters must not depend on each other:\n" + "\n".join(violations)
    )


def test_mock_v1_1_adapter_does_not_depend_on_v1_0():
    """The test-only mock v1.1 adapter in test_adapter_migration.py must
    also be self-contained — it models what a real adapter looks like
    (see docs/migrations/mock_v1_0_to_v1_1.md), so it should be held to the
    same standard, not just production code.
    """
    assert _MOCK_FILE.exists(), f"expected {_MOCK_FILE} to exist"
    violations = _violations_in_file(_MOCK_FILE, own_version="v1_1")
    assert not violations, (
        "Mock v1.1 adapter must not depend on v1.0:\n" + "\n".join(violations)
    )

File: cpp/tests/test_version_adapter_isolation.cpp
// Architectural guard: no File_Version adapter may import or call into
// another version's adapter code, even for logic that is byte-identical
// between versions today. If a later version depended on an earlier one's
// code, the earlier version could never be changed or removed without
// checking every dependent later version — and each additional version
// compounds the problem. This was already enforced for Python's real v1.1
// adapter and its test-only mock; this test enforces the same rule for
// C++'s version-scoped adapter files, including the test-only mock.
//
// Mechanism: read each relevant file's raw text and scan for substrings
// matching the shape `v<digits>_<digits>` (e.g. "v1_0", "v1_1"), case
// insensitively — this also catches per-version class names like
// "Hdf5AdapterV1_0" and "Hdf5WriterV1_0", not just namespace/include-path
// tokens. Each file's "own" version is derived from its location (the
// `vX_Y` directory it lives under, or — for the test-only mock, which has
// no directory of its own — its filename). Any *other* version token found
// in the file is a violation.
//
// Currently a no-op tripwire in the sense that it should already pass:
// only v1_0 is a real adapter today, and the mock at cpp/tests/mock_v1_1.hpp
// was made fully self-contained in the same change that added this test.
// It starts actually enforcing the moment a real v1_1 (or later) C++
// adapter is added and someone reaches back into v1_0's headers/namespace.
#include <catch2/catch_test_macros.hpp>

#include <filesystem>
#include <fstream>
#include <regex>
#include <sstream>
#include <string>
#include <vector>

#ifndef CPP_SOURCE_DIR
#  error "CPP_SOURCE_DIR must be defined by tests/CMakeLists.txt"
#endif

namespace {

namespace fs = std::filesystem;

std::string readFileText(const fs::path& p) {
    std::ifstream in(p, std::ios::binary);
    REQUIRE(in.is_open());
    std::ostringstream ss;
    ss << in.rdbuf();
    return ss.str();
}

// Every "v<digits>_<digits>" token in `text`, lower-cased, e.g. both
// "capabilities::v1_0" and "Hdf5AdapterV1_0" yield "v1_0".
std::vector<std::string> findVersionTokens(const std::string& text) {
    static const std::regex re(R"([vV](\d+)_(\d+))");
    std::vector<std::string> out;
    for (auto it = std::sregex_iterator(text.begin(), text.end(), re);
         it != std::sregex_iterator(); ++it) {
        out.push_back("v" + (*it)[1].str() + "_" + (*it)[2].str());
    }
    return out;
}

struct CheckedFile {
    fs::path path;
    std::string ownVersion;
};

void collectFilesUnder(const fs::path& dir, const std::string& ownVersion,
                        std::vector<CheckedFile>& out) {
    if (!fs::exists(dir)) return;
    for (const auto& entry : fs::recursive_directory_iterator(dir)) {
        if (entry.is_regular_file())
            out.push_back({entry.path(), ownVersion});
    }
}

std::vector<CheckedFile> collectCheckedFiles() {
    std::vector<CheckedFile> files;

    fs::path capsDir = fs::path(CPP_SOURCE_DIR) / "include/machine_config/capabilities";
    collectFilesUnder(capsDir / "v1_0", "v1_0", files);
    collectFilesUnder(capsDir / "v1_1", "v1_1", files); // no-op today; enforces once added

    fs::path mockPath = fs::path(CPP_SOURCE_DIR) / "tests/mock_v1_1.hpp";
    if (fs::exists(mockPath))
        files.push_back({mockPath, "v1_1"});

    return files;
}

} // namespace

TEST_CASE("version adapter isolation: no cross-version references") {
    auto files = collectCheckedFiles();

    // Sanity check: the guard isn't vacuously trivial — the real v1.0
    // adapter directory and the mock must both actually be found.
    bool sawV1_0 = false, sawMock = false;
    for (const auto& f : files) {
        if (f.ownVersion == "v1_0") sawV1_0 = true;
        if (f.path.filename() == "mock_v1_1.hpp") sawMock = true;
    }
    REQUIRE(sawV1_0);
    REQUIRE(sawMock);

    std::vector<std::string> violations;
    for (const auto& f : files) {
        std::string text = readFileText(f.path);
        for (const auto& token : findVersionTokens(text)) {
            if (token != f.ownVersion) {
                violations.push_back(
                    f.path.string() + ": found foreign version token '" + token +
                    "' (expected only '" + f.ownVersion + "')");
            }
        }
    }

    std::ostringstream msg;
    msg << "Cross-version references found:\n";
    for (const auto& v : violations) msg << "  " << v << "\n";
    INFO(msg.str());
    REQUIRE(violations.empty());
}
